Realm/ObjectStore/object_store_exceptions.cpp

```diff
@@ -17,29 +17,112 @@
+    m_info[InfoKeyPropertyType] = string_for_property_type(prop.type);
+    m_info[InfoKeyPropertyObjectType] = prop.object_type;
+    m_what = generate_what();
+}
+
+ObjectStoreException::ObjectStoreException(Kind kind, const std::string &object_type, const Property &prop, const Property &oldProp) :
+    m_kind(kind) {
+    m_info[InfoKeyObjectType] = object_type;
+    m_info[InfoKeyPropertyName] = prop.name;
+    m_info[InfoKeyPropertyType] = string_for_property_type(prop.type);
+    m_info[InfoKeyOldPropertyType] = string_for_property_type(oldProp.type);
+    m_info[InfoKeyPropertyObjectType] = prop.object_type;
+    m_info[InfoKeyOldPropertyObjectType] = oldProp.object_type;
+    m_what = generate_what();
+}
+
+ObjectStoreException::ObjectStoreException(Kind kind, const std::string &object_type, const std::string primary_key) : m_kind(kind) {
+    m_info[InfoKeyObjectType] = object_type;
+    m_info[InfoKeyPrimaryKey] = primary_key;
+    m_what = generate_what();
+}
+
+ObjectStoreException::ObjectStoreException(uint64_t old_version, uint64_t new_version) : m_kind(Kind::RealmVersionGreaterThanSchemaVersion) {
+    m_info[InfoKeyOldVersion] = to_string(old_version);
+    m_info[InfoKeyNewVersion] = to_string(new_version);
+    m_what = generate_what();
+}
+
+ObjectStoreException::ObjectStoreException(vector<ObjectStoreException> validation_errors, const string &object_type) :
+    m_validation_errors(validation_errors), m_kind(Kind::ObjectStoreValidationFailure) {
+    m_info[InfoKeyObjectType] = object_type;
+    m_what = generate_what();
+}
+
+string ObjectStoreException::generate_what() const {
+    auto format_string = s_custom_format_strings.find(m_kind);
+    if (format_string != s_custom_format_strings.end()) {
+        return populate_format_string(format_string->second);
     }
+    return populate_format_string(s_default_format_strings.at(m_kind));
 }
 
-ObjectStoreValidationException::ObjectStoreValidationException(std::vector<std::string> validation_errors, std::string object_type) :
-    m_validation_errors(validation_errors), m_object_type(object_type) {
-    m_what = "Migration is required for object type '" + m_object_type + "' due to the following errors:";
+string ObjectStoreException::validation_errors_string() const {
+    string errors_string;
     for (auto error : m_validation_errors) {
-        m_what += "\n- " + error;
+        errors_string += string("\n- ") + error.what();
     }
+    return errors_string;
 }
+
+std::string ObjectStoreException::populate_format_string(const std::string & format_string) const {
+    string out_string, current(format_string);
+    smatch sm;
+    regex re("\\{(\\w+)\\}");
+    while(regex_search(current, sm, re)) {
+        out_string += sm.prefix();
+        const string &key = sm[1];
+        if (key == "ValidationErrors") {
+            out_string += validation_errors_string();
+        }
+        else {
+            out_string += m_info.at(key);
```

Comment:
This throws an exception if `key` does not exist in `m_info`. I think the "correct" behavior in this case would be to actually leave the interpolation placeholder as is, so that specifying a key that doesn't exist "does nothing".

I.e., the format string "Hello, {KeyThatDoesNotExist}!" should simply yield the output "Hello, {KeyThatDoesNotExist}!", which is testable, but doesn't cause sudden crashes when we make a mistake.

Alternatively, all of this can be done away with if we switch to using actual specialized exception types that define members for all of this, in which case we would get a compile-time error instead. :)